Mathlib/MeasureTheory/Group/FundamentalDomain.lean

```diff
@@ -719,4 +727,315 @@ protected theorem fundamentalInterior : IsFundamentalDomain G (fundamentalInteri
+  which `IsFundamentalDomain G s` holds. -/
+class HasFundamentalDomain (G : Type*) (α : Type*) [One G] [SMul G α] [MeasureSpace α] : Prop :=
+(has_fundamental_domain_characterization : ∃ (s : Set α), IsFundamentalDomain G s)
+
+attribute [to_additive existing MeasureTheory.HasAddFundamentalDomain]
+  MeasureTheory.HasFundamentalDomain
+
+/-- The `covolume` of an action of `G` on `α` which has a fundamental domain is the volume of some
+fundamental domain. -/
+@[to_additive addCovolume "The `addCovolume` of an action of `G` on `α` which has a fundamental
+domain is the volume of some fundamental domain."]
+def covolume (G α : Type*) [One G] [SMul G α] [MeasureSpace α]
+    [funDom : HasFundamentalDomain G α] : ℝ≥0∞ :=
+  volume funDom.has_fundamental_domain_characterization.choose
+
+variable [Group G] [MulAction G α] [MeasurableSpace G] [MeasureSpace α]
+
+/-- If there is a fundamental domain `s`, then `HasFundamentalDomain` holds. -/
+@[to_additive]
+lemma IsFundamentalDomain.hasFundamentalDomain {s : Set α} (fund_dom_s : IsFundamentalDomain G s) :
+    HasFundamentalDomain G α := ⟨⟨s, fund_dom_s⟩⟩
+
+/-- The `covolume` can be computed by taking the `volume` of any given fundamental domain `s`. -/
+@[to_additive]
+lemma IsFundamentalDomain.covolume_eq_volume [Countable G] [MeasurableSMul G α]
+    [SMulInvariantMeasure G α volume] {s : Set α} (fund_dom_s : IsFundamentalDomain G s) :
+    covolume (funDom := fund_dom_s.hasFundamentalDomain) = volume s := by
+  dsimp [covolume]
+  rw [fund_dom_s.measure_eq]
+  exact fund_dom_s.hasFundamentalDomain.has_fundamental_domain_characterization.choose_spec
+
+end HasFundamentalDomain
+
+/-! ## `QuotientVolumeEqVolume` typeclass
+
+This typeclass describes a situation in which a measure `μ` on `α ⧸ G` can be computed by
+taking the measure on `α` of the intersection of the pullback with a fundamental domain.
+
+It's curious that in measure theory, measures can be pushed forward, while in geometry, volumes can
+be pulled back. And yet here, we are describing a situation involving measures in a geometric way.
+
+Another viewpoint is that if a set is small enough to fit in a single fundamental domain, then its
+measure in `α` is the same as the `μ` measure of its pushforward in `α ⧸ G`.
+
+-/
+
+section QuotientVolumeEqVolume
+
+section additive
+
+variable [AddGroup G] [AddAction G α] [MeasureSpace α]
+
+local notation "α_mod_G" => AddAction.orbitRel G α
+
+local notation "π" => @Quotient.mk _ α_mod_G
+
+/-- A measure `μ` on the `AddQuotient` of `α` mod `G` satisfies `AddQuotientVolumeEqVolumePreimage`
+if: for any fundamental domain `t`, and any measurable subset `U` of the quotient,
+`μ U = volume ((π ⁻¹' U) ∩ t)`. -/
+class AddQuotientVolumeEqVolumePreimage (μ : Measure (Quotient α_mod_G)) : Prop where
+add_projection_respects_measure' : ∀ (t : Set α) (_ : IsAddFundamentalDomain G t)
+  (U : Set (Quotient α_mod_G)) (_ : MeasurableSet U),
+  μ U = volume (π ⁻¹' U ∩ t)
```

Comment:
Is it better to write `μ = (volume.map π).restrict t`?

Reply:
Actually, the goal of this typeclass is to allow to work with sets upstairs; working with measures downstairs is more cumbersome than using a variety of corresponding sets upstairs. So we'd like to keep this as is.

Reply:
Am I right in saying that my spelling is equivalent? If so, could you add a lemma assuming `AddQuotientVolumeEqVolumePreimage` with my conclusion?

Reply:
Done, added `quotient_measure_eq_map_restrict`. Thanks!

Reply:
Whoops, it was already there! `IsFundamentalDomain.eq_quotientMeasure`